File: server/internal/handler/mcp_template.go
package handler

import (
	"net/http"

	"github.com/multica-ai/multica/server/internal/service"
)

// McpServerTemplateResponse is one entry in the built-in MCP catalog.
//
// Unlike WorkspaceMcpServerResponse (which is intentionally write-only and
// never carries url / command / args / headers / env), Config is served in
// full: these templates are public, credential-free content the workspace has
// not yet adopted, so there is nothing to protect. The keyless invariant is
// enforced by builtin_mcp_templates_test.go.
//
// No transport is sent: the client derives it from Config via mcpTransport, so
// echoing a server-computed value would just be a second source of truth.
type McpServerTemplateResponse struct {
	Key         string         `json:"key"`
	Title       string         `json:"title"`
	Description string         `json:"description"`
	Config      map[string]any `json:"config"`
}

// ListMcpServerTemplates returns the built-in MCP catalog in the requested
// language. Read-only and workspace-independent: templates ship with the
// binary, so this answers the same for every workspace. It lives behind the
// workspace-scoped API group so the client needs no special call shape, and it
// is member-visible for the same reason as the workspace MCP library — an agent
// owner needs to see what they can add.
func (h *Handler) ListMcpServerTemplates(w http.ResponseWriter, r *http.Request) {
	language := templateLanguageFromRequest(r.URL.Query().Get("language"))
	templates := service.McpServerTemplates()
	out := make([]McpServerTemplateResponse, 0, len(templates))
	for _, template := range templates {
		out = append(out, McpServerTemplateResponse{
			Key:         template.Key,
			Title:       template.Title(language),
			Description: template.Description(language),
			Config:      template.Config,
		})
	}
	writeJSON(w, http.StatusOK, map[string]any{"templates": out})
}

File: packages/core/types/mcp-template.ts
/**
 * A built-in MCP server template: one of the mainstream MCP servers a workspace
 * can add with one click.
 *
 * Selecting a template pre-fills the ordinary "add MCP server" form; saving
 * produces an ORDINARY workspace MCP server — write-only, renameable, and still
 * unassigned to any agent until an owner attaches it. Unlike the write-only
 * library, `config` is served in full because these templates are public,
 * credential-free content the workspace has not yet adopted.
 *
 * There is no `version` and no provenance stamp: the add form lets a user edit
 * the config before saving, so recording "came from template X" would be
 * dishonest. The transport is derived from `config` on the client, so it is not
 * a field here.
 */
export interface McpServerTemplate {
  /** Stable identity and default server name. Matches `^[A-Za-z0-9_-]+$`. */
  key: string;
  /** Localized catalog label. Follows the requested language. */
  title: string;
  /** Localized catalog description. */
  description: string;
  /** The raw MCP server entry copied verbatim into the add form. */
  config: Record<string, unknown>;
}
